### ISSUE TITLE
Is there any similar mechanism like arg_scope

### ISSUE THREAD

**Issue by mingyr** [deepmind/sonnet#92]
Dear Deepminders:

I just wonder is there a similar mechanism existing in the current sonnet repo resembling arg_scope?

Usually I would like to put all declarations of sub-modules in the initializer, but sometimes it looks a little redundant:

```
class Model(snt.AbstractModule):
    def __init__(self, name = "model"):
        super(Model, self).__init__(name = name)

        with self._enter_variable_scope():
            self._batch_norm_1 = BatchNorm(decay_rate = 0.9, update_ops_collection = None)
            self._batch_norm_2 = BatchNorm(decay_rate = 0.9, update_ops_collection = None)
            self._batch_norm_3 = BatchNorm(decay_rate = 0.9, update_ops_collection = None)
            self._batch_norm_4 = BatchNorm(decay_rate = 0.9, update_ops_collection = None)
            self._batch_norm_5 = BatchNorm(decay_rate = 0.9, update_ops_collection = None)
            self._batch_norm_6 = BatchNorm(decay_rate = 0.9, update_ops_collection = None)
            ......

```

I know I could do it in a list:
```
class Model(snt.AbstractModule):
    def __init__(self, name = "model"):
        super(Model, self).__init__(name = name)

        with self._enter_variable_scope():
            self._batch_norms = [BatchNorm(decay_rate = 0.9, update_ops_collection = None)
                                              for _ in range(num_batch_norm_modules)]
```

However, I still want to know if there is a way I can write as in the following:
```
class Model(snt.AbstractModule):
    def __init__(self, name = "model"):
        super(Model, self).__init__(name = name)

        with self._enter_variable_scope():
            with arg_scope([BatchNorm], decay_rate = 0.9, update_ops_collection = None): 
                self._batch_norm_1 = BatchNorm()
                self._batch_norm_2 = BatchNorm()
                self._batch_norm_3 = BatchNorm()
                self._batch_norm_4 = BatchNorm()
                self._batch_norm_5 = BatchNorm()
                self._batch_norm_6 = BatchNorm()
            .    .....

```

The arg_scope and its usage is from the following links:
https://github.com/tensorflow/models/blob/master/research/inception/inception/slim/scopes.py
https://github.com/tensorflow/models/blob/master/research/inception/inception/inception_model.py

I just put here for discussion since the current sonnet still works well.

**Comment by kosklain** [comment]
Hi,

There is not an exact analogy in Sonnet, but you can do exactly this with `batch_norm_creator = functools.partial(BatchNorm, decay_rate=0.9, ...)` and then `self._batch_norm_1 = batch_norm_creator()`  (`import functools`).

**Comment by mingyr** [comment]
The suggestion is nice. I will close the thread.
